packages/make-hardener/src/main.js

```diff
@@ -21,125 +21,125 @@
-
-        // Object are verified before being enqueued,
-        // therefore this is a valid candidate.
-        // Throws if this fails (strict mode).
-        freeze(obj);
-
-        // we rely upon certain commitments of Object.freeze and proxies here
-
-        // get stable/immutable outbound links before a Proxy has a chance to do
-        // something sneaky.
-        const path = paths.get(obj) || 'unknown';
-        const descs = getOwnPropertyDescriptors(obj);
-        const proto = getPrototypeOf(obj);
-        enqueue(proto, `${path}.__proto__`);
-
-        ownKeys(descs).forEach(name => {
-          const pathname = `${path}.${String(name)}`;
-          // todo uncurried form
-          // todo: getOwnPropertyDescriptors is guaranteed to return well-formed
-          // descriptors, but they still inherit from Object.prototype. If
-          // someone has poisoned Object.prototype to add 'value' or 'get'
-          // properties, then a simple 'if ("value" in desc)' or 'desc.value'
-          // test could be confused. We use hasOwnProperty to be sure about
-          // whether 'value' is present or not, which tells us for sure that
-          // this is a data property.
-          // The 'name' may be a symbol, and TypeScript doesn't like us to
-          // index arbitrary symbols on objects, so we pretend they're just
-          // strings.
-          const desc = descs[/** @type {string} */ (name)];
-          if ('value' in desc) {
-            // todo uncurried form
-            enqueue(desc.value, `${pathname}`);
-          } else {
-            enqueue(desc.get, `${pathname}(get)`);
-            enqueue(desc.set, `${pathname}(set)`);
-          }
-        });
-      }
-
-      function dequeue() {
-        // New values added before forEach() has finished will be visited.
-        toFreeze.forEach(freezeAndTraverse); // todo curried forEach
-      }
-
-      function commit() {
-        // todo curried forEach
-        // we capture the real WeakSet.prototype.add above, in case someone
-        // changes it. The two-argument form of forEach passes the second
-        // argument as the 'this' binding, so we add to the correct set.
-        toFreeze.forEach(hardened.add, hardened);
-      }
-
-      enqueue(root);
-      dequeue();
-      // console.log("toFreeze set:", toFreeze);
-      commit();
-
-      return root;
-    },
-  };
-
+    });
+  }
+
+  function dequeue() {
+    // New values added before forEach() has finished will be visited.
+    toHarden.forEach(freezeAndTraverse); // TODO curried forEach
+  }
+
+  enqueue(root, '<root>');
+  dequeue();
+  // The commit step. We must only record objects from `toHarden` as
+  // hardened if we record all of them as hardened.
+  moreObjectsHardened(toHarden);
+
+  return root;
+};
+
+let called = false;
+
+const makeHardener = () => {
+  if (called) {
+    throw new Error('makeHardener must only be called once');
+  }
+  called = true;
```

Comment:
Just to note, this is making the case stronger for moving make-hardener into SES-shim since it’s coupled to Lockdown, which we have already planned #41 but haven’t prioritized yet.